Sources/VectorExtor/CGPath+Extensions.swift

```diff
@@ -3,275 +3,152 @@ import CoreGraphics
 
 @available(OSX 10.13, iOS 11.0, tvOS 11.0, watchOS 4.0, *)
 public extension CGPath {
-	enum PathElement: Equatable {
-		case moveTo(point: CGPoint)
-		case addLineTo(point: CGPoint)
-		case addQuadCurveTo(point: CGPoint, controlPoint: CGPoint)
-		case addCurveTo(point: CGPoint, controlPoint1: CGPoint, controlPoint2: CGPoint)
-		case close
-	}
 
-	class PathSection {
-		public let element: PathElement
-		var next: PathSection?
-//		var previous: PathSection?
-		weak var previous: PathSection?
+	struct Segment: Hashable, Codable, Sendable {
+		public enum Curve: CurveProtocol, Hashable, Codable, Sendable {
+			case moveTo(MoveCurve)
+			case addLineTo(LineCurve)
+			case addQuadCurveTo(QuadCurve)
+			case addCurveTo(CubicCurve)
+			case close(CloseCurve)
+
+			public var length: Double {
+				switch self {
+				case
+					.addLineTo(let curve as CurveProtocol),
+					.addQuadCurveTo(let curve as CurveProtocol),
+					.addCurveTo(let curve as CurveProtocol),
+					.close(let curve as CurveProtocol),
+					.moveTo(let curve as CurveProtocol):
+					curve.length
+				}
+			}
 
-		var endPoint: CGPoint? {
-			switch element {
-			case .addLineTo(point: let point):
-				return point
-			case .addCurveTo(point: let point, controlPoint1: _, controlPoint2: _):
-				return point
-			case .addQuadCurveTo(point: let point, controlPoint: _):
-				return point
-			case .moveTo(point: let point):
-				return point
-			default:
-				return nil
+			public var startPoint: CGPoint? {
+				switch self {
+				case .moveTo(let curve):
+					curve.startPoint
+				case .addLineTo(let curve):
+					curve.startPoint
+				case .addQuadCurveTo(let curve):
+					curve.startPoint
+				case .addCurveTo(let curve):
+					curve.startPoint
+				case .close(let curve):
+					curve.startPoint
+				}
 			}
-		}
 
-		init(element: CGPath.PathElement, next: CGPath.PathSection? = nil, previous: CGPath.PathSection? = nil) {
-			self.element = element
-			self.next = next
-			self.previous = previous
+			public var endPoint: CGPoint {
+				switch self {
+				case .moveTo(let curve):
+					curve.endPoint
+				case .addLineTo(let curve):
+					curve.endPoint
+				case .addQuadCurveTo(let curve):
+					curve.endPoint
+				case .addCurveTo(let curve):
+					curve.endPoint
+				case .close(let curve):
+					curve.endPoint
+				}
+			}
 		}
+
+		public let curve: Curve
+		public var startingPoint: CGPoint? { curve.startPoint }
 	}
 
-	var elements: [PathElement] {
-		var store = [PathElement]()
+	var segments: [Segment] {
+		var accumulator: [Segment] = []
+		var closeEnd: CGPoint?
 		applyWithBlock { elementsPointer in
+			func setClose(_ new: CGPoint?) {
```

Comment:
perhaps setCloseEnd()?  as it stands, I'd call what it's doing a side effect even though that is clearly not your intent.

Reply:
Yeah - this part got a bit clunky. 

Each node in a path implicitly refers to the previous node's ending as the starting point for the new node. The model that CGPath uses with CGPathElement properly reflects that implicit nature. An additional advantage of this is that if a single node gets adjusted to a new location, it inherently updates the next node, as it (the next one) simply refers back to the moved one.

However, to do a segment length calculation (and dependent methods) I need to know not just the control and end points for the segment, but also the starting point, and a reference to the previous node frequently isn't available. This is actually where the deprecated code ran into the catch 22 with the pointer to `next` and `previous` being required, even for a single segment length calculation, which originally caused the memory leak. But when I made `previous` into a `weak` var, when you have a path, grab the last segment, then try to calculate the length of that curve alone, the `weak` nature of `previous` caused the prior node to deallocate, thereby eliminating our previous starting point. Thankfully, I had this scenario covered in my tests, cuz I wouldn't have noticed otherwise.

So, the refactor went with a model that duplicates the `endPoint` of one node with the `startPoint` of the next. That should be okay, because this isn't really meant as a means of modifying the path node by node, but even if it goes that direction, it's just a matter of updating both values.

Since all* nodes have a start and end point, I wanted to DRY up code with using a generic protocol, allowing for a normalized start and end on all nodes... *Except that `moveTo` and `closeSubpath` buck the trend. That meant that I had a few choices for how to deal with those...  
* I could make `startPoint` and `endPoint` optional, but that would make unwrapping their values a nuisance for the rest of the node types.
* I could make `startPoint` and `endPoint` equal to `.zero` where they aren't used, but that would technically be incorrect since the truth is that both `moveTo` and `closeSubpath` really have implicit values for both. For the most part, these implicit values are semi if not entirely irrelevant, but they *technically* exist. 
* Finally I could set them explicitly to their implicit values, which is the option I went with. This resulted in this clunky code, where I need to track the previous node's end for the start of the next node (which is really required nonetheless) AND track where the subpath begins to know where the `closeSubpath` would end.

How does `setTentativeCloseSubpathEnd` sound? I'll update it to that and see if you have any other ideas.